### ISSUE TITLE
Gene autocomplete service

### ISSUE THREAD

**Issue by alexpreynolds** [hms-dbmi/higlass#293]
Local installation at: http://hg-dev.altius.org/app/

_Commit_: 

Tag v0.6.31 of higlass-website
Master branch of higlass-server

#### Steps to reproduce

1. Toggle the position search box
2. Enter a portion or all of a gene name (CTCF, HBB, etc.)

#### Observed behavior

- No drop-down menu with gene selections is shown
- Browser console reports a 404 error from loading the following URL (for example):

```
http://higlass.io/api/v1/suggest/?d=ABXo2-QKSYyrdY35zoXZ4A&ac=hbb
```

The `ac` value looks to be the autocomplete request value (gene name being queried).

The `d` value `ABXo2-QKSYyrdY35zoXZ4A` appears to come from the `autocompleteId` property of a view object instance (for example):

```
      ...
      "uid": "MIbjEWf-SDSKrRINO3abKA",
      "genomePositionSearchBoxVisible": true,
      "genomePositionSearchBox": {
        "autocompleteServer": "http://higlass.io/api/v1",
        "chromInfoServer": "http://higlass.io/api/v1",
        "visible": true,
        "chromInfoId": "hg38",
        "autocompleteId": "ABXo2-QKSYyrdY35zoXZ4A"
      }
```

If I load the URL separately, I get the following JSON response:

```
{"detail":"Suggestion source file not found"}
```

#### Expected behavior

- Select menu is shown with gene suggestions

#### Workaround

If I delete the `uid` and `autocompleteId` properties of a view object, then autocomplete queries work again. However, this breaks viewport projections or linked views, which probably are relying on the `uid` values of views to apply zoom, drag, and other rendering events.

I have also tried changing the URLs of the `autocompleteServer` and `chromInfoServer` properties, but I didn't expect this to work without loading data into the local server (and it didn't work). 

Is there a way to set up a local autocomplete server?

**Comment by pkerpedjiev** [comment]
> Is there a way to set up a local autocomplete server?

Yes! You're actually already running one. Take a look here: http://hg-dev.altius.org/api/v1/suggest/?d=ABXo2-QKSYyrdY35zoXZ4A&ac=h

The `d=` parameter refers to the uuid of a gene annotations dataset on a server. So if you've created your own gene annotations dataset using the instructions here: http://docs.higlass.io/gene_annotations.html#creating-the-track, you can always get autocomplete suggestions using the `suggest` API.

**Comment by alexpreynolds** [comment]
Thanks, Peter. I did use those instructions, as it happens, just to get a feel for what input goes into a gene annotations track. 

Are there specific keys I should change in the configuration object, which will use our server? 

For instance, I have the following autocomplete settings for a view:

```
      ...
      "uid": "MIbjEWf-SDSKrRINO3abKA",
      "genomePositionSearchBoxVisible": true,
      "genomePositionSearchBox": {
        "autocompleteServer": "http://higlass.io/api/v1",
        "chromInfoServer": "http://higlass.io/api/v1",
        "visible": true,
        "chromInfoId": "hg38",
        "autocompleteId": "ABXo2-QKSYyrdY35zoXZ4A"
      }
```

I changed the `autocompleteServer` and `chromInfoServer` keys to:

```
      ...
      "uid": "MIbjEWf-SDSKrRINO3abKA",
      "genomePositionSearchBoxVisible": true,
      "genomePositionSearchBox": {
        "autocompleteServer": "http://hg-dev.altius.org/api/v1",
        "chromInfoServer": "http://hg-dev.altius.org/api/v1",
        "visible": true,
        "chromInfoId": "hg38",
        "autocompleteId": "ABXo2-QKSYyrdY35zoXZ4A"
      }
```

It looks like our hg-dev site still sends requests to `higlass.io` for autocomplete suggestions, with the resulting 404 error. Are there other adjustments required?

**Comment by alexpreynolds** [comment]
For some reason, a view's `state` includes the `autocompleteServer` and `autocompleteId` properties, but they do not seem to be read in from the JSON object being loaded in via the initial `launchHg` function call (or used), and instead these two properties seem to be read from default settings elsewhere in `hglib.js`.

I found that editing lines L73803 and L73676 of `hglib.js` to manually replace `this.state.autocompleteServer` with the string `"http://hg-dev.altius.org/api/v1"` and `this.state.autocompleteId` with the string `"XcTBzvqBRO6HbnufF2b1ZQ"` seems to work.

It's just odd that these two properties are not being processed from the JSON object I'm using to configure the initial views.

If it is useful, here is the JSON object I am using to configure the initial view:

```
wget -O- --user encode --password collabor8 
 https://resources.altius.org/~areynolds/higlass.config.json.txt > /tmp/higlass.config.json.txt
```

**Comment by pkerpedjiev** [comment]
This is a bug. I just pushed a commit that fixes it which will be out in the next release. I'll close the issue once it's released.

Let me know if this is urgent and I can try to speed things up and get a release out earlier. Otherwise I'll try to lump this with some other bug fixes.

**Comment by pkerpedjiev** [comment]
On a side note, I think there's something wrong with your gene annotations. Autocomplete returns results like this:

![image](https://user-images.githubusercontent.com/2143629/43731409-6849f4c2-997c-11e8-9a30-7c2a11b65ae0.png)

It might have something to do with this: https://github.com/hms-dbmi/higlass/issues/294

**Comment by pkerpedjiev** [comment]
Oh, never mind, those really are your "gene" names :)

**Comment by alexpreynolds** [comment]
Thanks, this would be a low priority issue for me, as I can fix it in the JSON object.
